src/irmin-pack/unix/control_file.ml

```diff
@@ -299,35 +299,43 @@ module Make (Serde : Serde.S) (Io : Io.S) = struct
   module Io = Io
 
   type payload = Serde.payload
-  type t = { io : Io.t; mutable payload : payload }
+
+  type t = {
+    mutable io : Io.t;
+    mutable payload : payload;
+    path : string;
+    tmp_path : string;
+  }
 
   let write io payload =
     let s = Serde.to_bin_string payload in
-    (* The data must fit inside a single page for atomic updates of the file.
-       This is only true for some file systems. This system will have to be
-       reworked for [V4]. *)
-    assert (String.length s <= Io.page_size);
-
     Io.write_string io ~off:Int63.zero s
 
+  let set_payload t payload =
+    let open Result_syntax in
+    let* () = Io.close t.io in
+    let* io_tmp = Io.create ~path:t.tmp_path ~overwrite:true in
+    t.io <- io_tmp;
+    let* () = write io_tmp payload in
+    let+ () = Io.move_file ~src:t.tmp_path ~dst:t.path in
+    t.payload <- payload
+
   let read io =
     let open Result_syntax in
     let* string = Io.read_all_to_string io in
-    (* Since the control file is expected to fit in a page,
-       [read_all_to_string] should be atomic for most filesystems. *)
     Serde.of_bin_string string
 
-  let create_rw ~path ~overwrite (payload : payload) =
+  let create_rw ~path ~tmp_path ~overwrite (payload : payload) =
     let open Result_syntax in
     let* io = Io.create ~path ~overwrite in
     let+ () = write io payload in
-    { io; payload }
+    { io; payload; path; tmp_path }
 
-  let open_ ~path ~readonly =
+  let open_ ~path ~tmp_path ~readonly =
```

Comment:
Not really needed for this PR, but from an API perspective, it's a bit of a bummer that a caller has to specify `tmp_path` even when it is not needed for `readonly:true`.

Reply:
I actually though about this case but could not think of a better way to do it except by using an option, but it would add some error cases if not provided in a read write instance.

My personal choice would be to totally forbid the use of `open_` for readonly instances, and force them to use `read_payload` instead, as there are no reasons to keep an open Io for them (it has to be reopened everytime anyway)

Reply:
Actually we kinda need to keep the io opened for calls to fsync from the file manager.
If we were to remove the io, we might need to call fsync every time to ensure we have properly written the data.
It might come as a huge cost in performances

Reply:
Mm, mentioning `fsync` gave me a moment to think about the change in this PR a little more.

In practice, `fsync` isn't used (since `Fm.t.use_fsync = false` for octez), but if we wanted to most properly support `fsync` with this temporary file + rename atomic trick, we would need to change a little how the code works.

To quote https://lwn.net/Articles/457667/
>  Similarly, if you encounter a system failure (such as power loss, ENOSPC or an I/O error) while overwriting a file, it can result in the loss of existing data. To avoid this problem, it is common practice (and advisable) to write the updated data to a temporary file, ensure that it is safe on stable storage, then rename the temporary file to the original file name (thus replacing the contents). This ensures an atomic update of the file, so that other readers get one copy of the data or another. The following steps are required to perform this type of update:
>
>    1. create a new temp file (on the same file system!)
>    2. write data to the temp file
>    3. fsync() the temp file
>    4. rename the temp file to the appropriate name
>    5. fsync() the containing directory 

Currently, we would only call `fsync` on the new temporary file (not the original io, which is why I think it could be closed) after it has been renamed. To do the above, we would need to change the code to accept a `use_fsync` flag so that internally we could do this. We would also need to merge #2132. It was opened last year to add `fsync` support for directories, but it kind of stalled.

--

FWIW, I don't think this is an urgent change (again, `use_fsync` is false in practice) but something worth highlighting.

Reply:
I see, thanks a lot for bringing that up, it seems like the best option but as you pointed out, we ignore `fsync` in octez, once the PR is merged I'll create an issue to keep track of this for a future modification